### DISCUSSION TITLE
Is it possible to synchronously call an async function in the webloop?

### DISCUSSION THREAD

**Discussion by CNSeniorious000** [pyodide/pyodide#4670] [Q&A] [answered]
Like #2219, I wonder whether there is a replacement for `asyncio.run` to synchronously run an async function.

Normally, `asyncio.run_until_complete` should not returns a future, but in pyodide it does. Is there any way to do this? By context switching?By polling with `Atomic`?

**Comment by hoodmane** [comment, marked as answer]
With JavaScript Promise Integration you can do it. You'll need to opt into the chrome origin trial for jspi and also I'd suggest waiting for the next Pyodide release since there's several important bug fixes for the stack switching logic coming in Pyodide v0.26.

**Reply by tom-dudley** [reply]
Adding a bit of info on this which @hoodmane provided in Discord:

> - To enable just on your chromium family browser, go to `chrome://flags` and set `Experimental WebAssembly JavaScript Promise Integration (JSPI)` to `enabled`.
>  - To enable for a domain, you have to register the domain [here](https://developer.chrome.com/origintrials/#/view_trial/1603844417297317889) and I think you need to either add a header or a `<meta>` tag to enable it for each page.
>  - If you do this, you can pass `enableRunUntilComplete: true` to [loadPyodide](https://pyodide.org/en/stable/usage/api/js-api.html#globalThis.loadPyodide) and then `run_until_complete` will work as expected if it can.
> - You need to ensure also that the Python execution context is entered either via `pyodide.runPythonAsync` or by calling an async Python function or by calling a sync function with [pyFunc.callPromising](https://pyodide.org/en/stable/usage/api/js-api.html#pyodide.ffi.PyCallable.callPromising) otherwise stack switching will fail with a runtime error.

Note: The JSPI proposal can be found here: https://github.com/WebAssembly/js-promise-integration
